# Laboratorio 2

Se trabajará con el dataset de permisos de edificación de San Francisco:
https://www.kaggle.com/aparnashastry/building-permit-applications-data/data

In [37]:
import pandas as pd
import numpy as np
np.random.seed(32423)

## Importacion de datos

In [38]:
dateparse = lambda c: pd.to_datetime(c, format='%m/%d/%Y', errors='coerce')
col_dates = ['Permit Creation Date', 'Current Status Date', 'Filed Date', 'Issued Date', 'Completed Date','First Construction Document Date', 'Permit Expiration Date']
data = pd.read_csv("Building_Permits.csv", low_memory=False, parse_dates=col_dates, date_parser=dateparse, \
                   index_col=['Record ID'])

Vemos si los tipos de datos de las columnas fueron inferidos correctamente según la información suministrada en
https://www.kaggle.com/aparnashastry/building-permit-applications-data/data

In [39]:
data.dtypes

Permit Number                                     object
Permit Type                                        int64
Permit Type Definition                            object
Permit Creation Date                      datetime64[ns]
Block                                             object
Lot                                               object
Street Number                                      int64
Street Number Suffix                              object
Street Name                                       object
Street Suffix                                     object
Unit                                             float64
Unit Suffix                                       object
Description                                       object
Current Status                                    object
Current Status Date                       datetime64[ns]
Filed Date                                datetime64[ns]
Issued Date                               datetime64[ns]
Completed Date                 

### Fechas

Vemos cómo se parsearon las fechas nulas:

In [40]:
data['Completed Date'][:5]

Record ID
1380611233945          NaT
1420164406718          NaT
1424856504716          NaT
1443574295566   2017-07-24
144548169992           NaT
Name: Completed Date, dtype: datetime64[ns]

In [41]:
data.describe(include='all')

Permit Number    Permit Type  Permit Type Definition  \
count         198900  198900.000000                  198900   
unique        181495            NaN                       8   
top     201602179765            NaN  otc alterations permit   
freq             101            NaN                  178844   
first            NaN            NaN                     NaN   
last             NaN            NaN                     NaN   
mean             NaN       7.522323                     NaN   
std              NaN       1.457451                     NaN   
min              NaN       1.000000                     NaN   
25%              NaN       8.000000                     NaN   
50%              NaN       8.000000                     NaN   
75%              NaN       8.000000                     NaN   
max              NaN       8.000000                     NaN   

       Permit Creation Date   Block     Lot  Street Number  \
count                198900  198900  198900  198900.000000   
unique                 1291    4893    1050            NaN   
top     2017-09-15 00:00:00    3708       1            NaN   
freq                    413    1195   10114            NaN   
first   2012-03-28 00:00:00     NaN     NaN            NaN   
last    2018-02-23 00:00:00     NaN     NaN            NaN   
mean                    NaN     NaN     NaN    1121.728944   
std                     NaN     NaN     NaN    1135.768948   
min                     NaN     NaN     NaN       0.000000   
25%                     NaN     NaN     NaN     235.000000   
50%                     NaN     NaN     NaN     710.000000   
75%                     NaN     NaN     NaN    1700.000000   
max                     NaN     NaN     NaN    8400.000000   

       Street Number Suffix Street Name Street Suffix  \
count                  2216      198900        196132   
unique                   18        1704            21   
top                       A      Market            St   
freq                   1501        5443        138358   
first                   NaN         NaN           NaN   
last                    NaN         NaN           NaN   
mean                    NaN         NaN           NaN   
std                     NaN         NaN           NaN   
min                     NaN         NaN           NaN   
25%                     NaN         NaN           NaN   
50%                     NaN         NaN           NaN   
75%                     NaN         NaN           NaN   
max                     NaN         NaN           NaN   

                         ...                     TIDF Compliance  \
count                    ...                                   2   
unique                   ...                                   2   
top                      ...                                   Y   
freq                     ...                                   1   
first                    ...                                 NaN   
last                     ...                                 NaN   
mean                     ...                                 NaN   
std                      ...                                 NaN   
min                      ...                                 NaN   
25%                      ...                                 NaN   
50%                      ...                                 NaN   
75%                      ...                                 NaN   
max                      ...                                 NaN   

       Existing Construction Type Existing Construction Type Description  \
count               155534.000000                                 155534   
unique                        NaN                                      5   
top                           NaN                         wood frame (5)   
freq                          NaN                                 113350   
first                         NaN                                    NaN   
last                          NaN              

Ahora deberíamos poder calcular el rango de fechas para cada columna de tipo date

In [42]:
for date_type_col in col_dates:
    print(data[date_type_col].max() - data[date_type_col].min())

2158 days 00:00:00
2250 days 00:00:00
1878 days 00:00:00
1878 days 00:00:00
1876 days 00:00:00
1878 days 00:00:00
3873 days 00:00:00


## Asegurar de tener ids/claves únicas

Vemos si hay IDs duplicados

In [43]:
data[data.index.duplicated()]

Empty DataFrame
Columns: [Permit Number, Permit Type, Permit Type Definition, Permit Creation Date, Block, Lot, Street Number, Street Number Suffix, Street Name, Street Suffix, Unit, Unit Suffix, Description, Current Status, Current Status Date, Filed Date, Issued Date, Completed Date, First Construction Document Date, Structural Notification, Number of Existing Stories, Number of Proposed Stories, Voluntary Soft-Story Retrofit, Fire Only Permit, Permit Expiration Date, Estimated Cost, Revised Cost, Existing Use, Existing Units, Proposed Use, Proposed Units, Plansets, TIDF Compliance, Existing Construction Type, Existing Construction Type Description, Proposed Construction Type, Proposed Construction Type Description, Site Permit, Supervisor District, Neighborhoods - Analysis Boundaries, Zipcode, Location]
Index: []

[0 rows x 42 columns]

Ver si hay filas duplicadas y descartarlas

In [44]:
data[data.duplicated()]

Permit Number  Permit Type            Permit Type Definition  \
Record ID                                                                    
1294939199934  201302019267            8            otc alterations permit   
1300835395703  201303283290            8            otc alterations permit   
1301529164635  201304124392            8            otc alterations permit   
1308830100339  201306210141            2       new construction wood frame   
1316521206878  201309056009            8            otc alterations permit   
1317158490355  201309116496            8            otc alterations permit   
1319970206124  201310078627            8            otc alterations permit   
1319981387943  201309045902            8            otc alterations permit   
1320387100653  201310108954            8            otc alterations permit   
1320438118710  201310108990            8            otc alterations permit   
1321690416908  201310229954            8            otc alterations permit   
1321961102273  201310240168            8            otc alterations permit   
132438569891        M444587            8            otc alterations permit   
1324681170997       M445387            8            otc alterations permit   
1325247493716  201311252729            8            otc alterations permit   
132623099137   201312063490            8            otc alterations permit   
132774263349   201312204701            8            otc alterations permit   
1334249465250       M467927            8            otc alterations permit   
1335255436500  201403120593            8            otc alterations permit   
1337412393485  201402057839            8            otc alterations permit   
1342196420228  201405165994            8            otc alterations permit   
1342252496263       M489027            8            otc alterations permit   
134303490330   201404153296            8            otc alterations permit   
134870373331        M503928            8            otc alterations permit   
1349370392081  201407181601            8            otc alterations permit   
1350144289258  201407252216            8            otc alterations permit   
1356614144941       M520487            8            otc alterations permit   
135887277085        M527287            8            otc alterations permit   
1362652404491       M537887            8            otc alterations permit   
1364264394477  201412042999            8            otc alterations permit   
...                     ...          ...                               ...   
1406565249051       M646687            8            otc alterations permit   
1407193164635  201512215598            3  additions alterations or repairs   
1411187159237  201602028646            8            otc alterations permit   
141370258893   201512175353            8            otc alterations permit   
1423046419304  201605127286            8            otc alterations permit   
142405696527   201605208035            8            otc alterations permit   
1427347504407       M699308            8            otc alterations permit   
1433589450534  201608124949            3  additions alterations or repairs   
1448640491307  201612275918            1                  new construction   
1454408508529  201702270179            8            otc alterations permit   
1455163266268  201702099112            8            otc alterations permit   
145516457812   201702099112            8            otc alterations permit   
1460314165260       M779929            8            otc alterations permit   
1460729231465  201702099056            3  additions alterations or repairs   
1461223125343  201612144960            3  additions alterations or repairs   
1463991501743  201612215594            3  additions alterations or repairs   
1466273273818  201706088863            3  additions alterations or repairs   
147236396498        M814467            8            otc alterations permit   
1472719133304  201708013471         

In [45]:
data = data.drop_duplicates()
data[data.duplicated()]

Empty DataFrame
Columns: [Permit Number, Permit Type, Permit Type Definition, Permit Creation Date, Block, Lot, Street Number, Street Number Suffix, Street Name, Street Suffix, Unit, Unit Suffix, Description, Current Status, Current Status Date, Filed Date, Issued Date, Completed Date, First Construction Document Date, Structural Notification, Number of Existing Stories, Number of Proposed Stories, Voluntary Soft-Story Retrofit, Fire Only Permit, Permit Expiration Date, Estimated Cost, Revised Cost, Existing Use, Existing Units, Proposed Use, Proposed Units, Plansets, TIDF Compliance, Existing Construction Type, Existing Construction Type Description, Proposed Construction Type, Proposed Construction Type Description, Site Permit, Supervisor District, Neighborhoods - Analysis Boundaries, Zipcode, Location]
Index: []

[0 rows x 42 columns]

## Caracteres especiales en nombres de las columnas

In [46]:
import chardet
# look at the first ten thousand bytes to guess the character encoding
with open("Building_Permits.csv", 'rb') as rawdata:
    result = chardet.detect(rawdata.read(10000))

# check what the character encoding might be
print(result)

{'confidence': 1.0, 'encoding': 'ascii', 'language': ''}


El encoding resulta ser ASCII. Chequeamos que no hayan caracteres fuera de la A-Z y 0-9:

In [47]:
cols_special_chars = data.columns[data.columns != data.columns.str.extract(r'^(\w+)$')] 
cols_special_chars

/Users/andres/miniconda3/envs/dsaml/lib/python3.5/site-packages/ipykernel_launcher.py:1: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  """Entry point for launching an IPython kernel.


Index(['Permit Number', 'Permit Type', 'Permit Type Definition',
       'Permit Creation Date', 'Street Number', 'Street Number Suffix',
       'Street Name', 'Street Suffix', 'Unit Suffix', 'Current Status',
       'Current Status Date', 'Filed Date', 'Issued Date', 'Completed Date',
       'First Construction Document Date', 'Structural Notification',
       'Number of Existing Stories', 'Number of Proposed Stories',
       'Voluntary Soft-Story Retrofit', 'Fire Only Permit',
       'Permit Expiration Date', 'Estimated Cost', 'Revised Cost',
       'Existing Use', 'Existing Units', 'Proposed Use', 'Proposed Units',
       'TIDF Compliance', 'Existing Construction Type',
       'Existing Construction Type Description', 'Proposed Construction Type',
       'Proposed Construction Type Description', 'Site Permit',
       'Supervisor District', 'Neighborhoods - Analysis Boundaries'],
      dtype='object')

In [48]:
data.columns = data.columns.str.replace(' ', '_')
data.columns = data.columns.str.replace('[^\w]+', '')
data.columns = data.columns.str.lower()
data.columns

Index(['permit_number', 'permit_type', 'permit_type_definition',
       'permit_creation_date', 'block', 'lot', 'street_number',
       'street_number_suffix', 'street_name', 'street_suffix', 'unit',
       'unit_suffix', 'description', 'current_status', 'current_status_date',
       'filed_date', 'issued_date', 'completed_date',
       'first_construction_document_date', 'structural_notification',
       'number_of_existing_stories', 'number_of_proposed_stories',
       'voluntary_softstory_retrofit', 'fire_only_permit',
       'permit_expiration_date', 'estimated_cost', 'revised_cost',
       'existing_use', 'existing_units', 'proposed_use', 'proposed_units',
       'plansets', 'tidf_compliance', 'existing_construction_type',
       'existing_construction_type_description', 'proposed_construction_type',
       'proposed_construction_type_description', 'site_permit',
       'supervisor_district', 'neighborhoods__analysis_boundaries', 'zipcode',
       'location'],
      dtype='object'

## Valores faltantes

Veamos el % de filas las cuales tienen al menos un dato nulo:

In [49]:
len(data.dropna(axis=1)) / len(data) * 100

100.0

Claramente, tirar las filas con datos nulos no es una opción, ya que todas tienen al menos una columna nula.

### Rellenar con algún valor los campos con datos faltantes

Analizamos el % de datos faltantes en cada columna

In [50]:
missing_values_count = data.isna().sum()
(missing_values_count[missing_values_count > 0] / len(data)).to_frame()

0
street_number_suffix                    0.988865
street_suffix                           0.013916
unit                                    0.851789
unit_suffix                             0.990142
description                             0.001448
issued_date                             0.075079
completed_date                          0.511339
first_construction_document_date        0.075109
structural_notification                 0.965192
number_of_existing_stories              0.215072
number_of_proposed_stories              0.215505
voluntary_softstory_retrofit            0.999824
fire_only_permit                        0.905347
permit_expiration_date                  0.260784
estimated_cost                          0.191364
revised_cost                            0.030478
existing_use                            0.206678
existing_units                          0.259094
proposed_use                            0.213352
proposed_units                          0.255951
plansets                                0.187556
tidf_compliance                         0.999990
existing_construction_type              0.217999
existing_construction_type_description  0.217999
proposed_construction_type              0.216983
proposed_construction_type_description  0.216983
site_permit                             0.973068
supervisor_district                     0.008630
neighborhoods__analysis_boundaries      0.008671
zipcode                                 0.008625
location                                0.008545

Antes de descartar los valores nulos, analicemos las columnas a las que le faltan un % elevado de datos (más del 50%).
Nos valemos de la información adicional provista en https://www.kaggle.com/aparnashastry/building-permit-applications-data/data donde podemos encontar una breve descripción sobre cada columna.

1. street_number_suffix: es NULL el 98,86% de las filas, lo cual tiene sentido porque se trata
de una información complementaria al número de la calle que se utiliza en casos muy puntuales (street_number por su
parte no figura siquiera entre las columnas con algún dato faltante).
2. unit: está documentado como "unit of a building". No está claro si el valor es optativo; la alta cantidad de filas de este tipo sugeriría que sí.
3. unit_suffix: Información adicional de la columna anterior; puede ser null.
4. completed_data: Esta columna puede ser nula si no se terminó de edificar aún.

Ahora analizamos una serie de columnas en las que se indica si la edificación posee o no cierta cualidad. A primera vista, parece que se ha indicado únicamente las edificaciones que poseen esa cualidad con 'Y' y se ha dejado en blanco aquellas que no lo hacen. En tal caso, se podría rellenar esas columnas con datos faltantes con 'N' (o directamente binarizar el campo y marcar con '0' y '1'). Veamos si esto es así:

In [51]:
print('structural_notification\n' + str(data.structural_notification.value_counts()))
print('voluntary_softstory_retrofit\n' + str(data.voluntary_softstory_retrofit.value_counts()))
print('fire_only_permit\n' + str(data.fire_only_permit.value_counts()))
print('tidf_compliance\n' + str(data.tidf_compliance.value_counts()))
print('site_permit\n' + str(data.site_permit.value_counts()))

structural_notification
Y    6921
Name: structural_notification, dtype: int64
voluntary_softstory_retrofit
Y    35
Name: voluntary_softstory_retrofit, dtype: int64
fire_only_permit
Y    18820
Name: fire_only_permit, dtype: int64
tidf_compliance
Y    1
P    1
Name: tidf_compliance, dtype: int64
site_permit
Y    5355
Name: site_permit, dtype: int64


La presunción era cierta, salvo para la columna **tidf_compliance**. La documentación dice textual:
*"TIDF compliant or not"* lo cual indicaría que debería ser un campo tipo *booleano*. Aquí haremos una suposición
e intepretaremos a la letra P como "parcial", ya que se trata del cumplimiento o no de una normativa o estándar.

In [52]:
fill_cols = ['tidf_compliance', 'structural_notification', 'voluntary_softstory_retrofit', 'fire_only_permit', 'site_permit']
for col in fill_cols:
    data[col] = data[col].fillna(value='N')


## Codificación de variables

### Variable *location*

La columna *location* tiene la latitud y longitud de la edificación, que deberá ser separada en dos columnas distintas
para su correcta utilización

In [53]:
data['latitude'], data['longitude'] = data['location'].str.split(',', 1).str
data.latitude = data.latitude.str.replace('(','')
data.longitude = data.longitude.str.replace(')','')
data.latitude = data.latitude.astype(float)
data.longitude = data.longitude.astype(float)
data = data.drop(['location'], 1)

In [54]:
data.longitude.sample(3)

Record ID
1491627400578   -122.433250
1468075364100   -122.430844
1430521183513   -122.413018
Name: longitude, dtype: float64

### Variables categóricas

Serán variables categóricas todas aquellas cuyo tipo es string, y no es un campo descriptivo (i.e. con texto libre).


In [55]:
from sklearn import preprocessing

categoric_vars = ['lot', 'permit_type', 'permit_type_definition', 'street_number_suffix', 'street_name', 'street_suffix', 'unit_suffix', \
                  'current_status', 'structural_notification', 'voluntary_softstory_retrofit', 'fire_only_permit', \
                  'existing_use', 'proposed_use', 'tidf_compliance', 'existing_construction_type_description', \
                  'proposed_construction_type_description', 'neighborhoods__analysis_boundaries' \
                 ]

En el caso de *neighborhoods__analysis_boundaries* parece que tenemos filas con más de una categoría.


In [56]:
data['neighborhoods__analysis_boundaries'].value_counts()

Financial District/South Beach    21808
Mission                           14678
Sunset/Parkside                   10204
West of Twin Peaks                 8737
Castro/Upper Market                8527
Pacific Heights                    8505
Marina                             8241
Outer Richmond                     7850
Noe Valley                         7842
South of Market                    7565
Bernal Heights                     6066
Nob Hill                           6009
Haight Ashbury                     5796
Inner Sunset                       5774
Bayview Hunters Point              5665
Russian Hill                       5495
Hayes Valley                       5488
Tenderloin                         4781
Inner Richmond                     4453
Potrero Hill                       4289
Presidio Heights                   4083
North Beach                        4050
Western Addition                   3866
Chinatown                          3764
Lone Mountain/USF                  3356


En realidad, vemos que el caso de 'Castro/Upper Market' por tomar un ejemplo no es una doble categoría ya que no se observan ocurrencias de 'Castro' o 'Upper Market', por lo tanto podemos tomarlo (este y demás casos) como una sola catergoría.

Hacemos lo mismo con las demás:

In [57]:
for col in categoric_vars:
    print(data[col].value_counts())
    print("\n")

1       10111
7        5316
2        5182
3        5042
6        4833
8        4773
9        4590
5        4546
4        4381
11       4235
10       4159
12       4110
21       4089
14       4048
13       4032
16       3979
17       3913
20       3898
19       3863
18       3813
15       3734
22       3479
23       3390
26       3263
25       3125
28       2985
24       2936
32       2734
29       2659
27       2653
        ...  
847         1
050A        1
450         1
430         1
073C        1
17A         1
017H        1
504         1
582         1
518         1
601         1
018Y        1
387         1
043D        1
036D        1
300A        1
072A        1
072C        1
006N        1
371         1
013K        1
400         1
655         1
001Y        1
330         1
967         1
471         1
014Q        1
460         1
703         1
Name: lot, Length: 1050, dtype: int64


8    178789
3     14651
4      2892
2       949
6       600
7       511
1       348
5        91
Name: perm

Financial District/South Beach    21808
Mission                           14678
Sunset/Parkside                   10204
West of Twin Peaks                 8737
Castro/Upper Market                8527
Pacific Heights                    8505
Marina                             8241
Outer Richmond                     7850
Noe Valley                         7842
South of Market                    7565
Bernal Heights                     6066
Nob Hill                           6009
Haight Ashbury                     5796
Inner Sunset                       5774
Bayview Hunters Point              5665
Russian Hill                       5495
Hayes Valley                       5488
Tenderloin                         4781
Inner Richmond                     4453
Potrero Hill                       4289
Presidio Heights                   4083
North Beach                        4050
Western Addition                   3866
Chinatown                          3764
Lone Mountain/USF                  3356


Procedemos a transformar a números estas categorías

#### Variables con  orden total

Convertimos en binario los valores de las 4 columnas a las que rellenamos los valores faltantes con 'N'.

In [58]:
from sklearn.preprocessing import LabelBinarizer

bin_vars = ['voluntary_softstory_retrofit', 'structural_notification', 'fire_only_permit', 'site_permit']
ord_vars = bin_vars + ['tidf_compliance']

for bin_col in bin_vars:
    lb = LabelBinarizer()
    lb_results = lb.fit_transform(data[bin_col])
    data[bin_col] = lb_results
    print(data[bin_col].value_counts())

0    198796
1        35
Name: voluntary_softstory_retrofit, dtype: int64
0    191910
1      6921
Name: structural_notification, dtype: int64
0    180011
1     18820
Name: fire_only_permit, dtype: int64
0    193476
1      5355
Name: site_permit, dtype: int64


Para *tidf_compliance* usaremos otro convertidor ya que hay 3 valores posibles.
Hay que procurar que el valor 'P' sea mapeado a 1, 'Y' a 2 y 'N' a 0.

In [59]:
P_index = data[data.tidf_compliance == 'P'].index
Y_index = data[data.tidf_compliance == 'Y'].index

In [60]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
le.fit(['N','P','Y'])
data.tidf_compliance = le.transform(data.tidf_compliance)
data['tidf_compliance'].value_counts()

0    198829
2         1
1         1
Name: tidf_compliance, dtype: int64

Corroboramos que la 'P' haya sido mapeada al 1

In [61]:
print(data[data.tidf_compliance == 2].index == Y_index)
print(data[data.tidf_compliance == 1].index == P_index)

[ True]
[ True]


#### Variables con valores no ordinales

Convertimos las variables que no deberían ser ordinales extrayendo cada valor posible en una nueva columna binaria. 
Para ello usamos el método 'get_dummies' de pandas en lugar de los encoders de sklearn

In [62]:
many_cat = [var for var in categoric_vars if var not in ord_vars]
dummies = pd.get_dummies(data, columns=many_cat, prefix=many_cat)
print(data.current_status.value_counts())
print(dummies.shape)
data = pd.concat([dummies, data])
print(data.shape)

complete       97047
issued         83532
filed          12033
withdrawn       1753
cancelled       1535
expired         1370
approved         733
reinstated       563
suspend          193
revoked           50
plancheck         16
appeal             2
incomplete         2
disapproved        2
Name: current_status, dtype: int64
(198831, 3255)
(397662, 3268)


## Consistencia de las variables

Chequeamos que las fechas cargadas en cada fila sean consistentes entre sí. Por ejemplo, 'Permit Creation Date' no puede ser anterior a 'Filed Date' (lo dice la propia documentación).
Los campos a revisar son entonces 'Filed Date', 'Permit Creation Date', 'Current Status Date' y 'Completed Date'.

In [63]:
(data['filed_date'] <= data['permit_creation_date']).value_counts()

True     396678
False       984
dtype: int64

Veamos las filas en las que las fechas son insconsistentes para corroborar tal hecho

In [64]:
data[data['filed_date'] > data['permit_creation_date']]
samples = data.sample(10)
print(samples['filed_date'])
print(samples['permit_creation_date'])

Record ID
1307439255952   2013-06-10
143405797435    2016-08-17
1476434136457   2017-08-28
1405685134580   2015-12-08
1481935487821   2017-10-03
147950492143    2017-09-15
1380456293356   2015-05-05
1465192464371   2017-06-01
1363699168319   2014-11-26
1326833165424   2013-12-11
Name: filed_date, dtype: datetime64[ns]
Record ID
1307439255952   2013-06-10
143405797435    2016-08-17
1476434136457   2017-08-28
1405685134580   2015-12-08
1481935487821   2017-10-03
147950492143    2017-09-15
1380456293356   2015-05-05
1465192464371   2017-06-01
1363699168319   2014-11-26
1326833165424   2013-12-11
Name: permit_creation_date, dtype: datetime64[ns]


Vemos algunos casos en los que supuestamente hay inconsistencia, pero se trata en ambos casos de la misma fecha, lo cual según la documentación es legal. El falso positivo se debería a un problema con la comparación entre datetimes.

Otra regla que hay que revisar si se cumple es que las filas con el campo 'Completed Date' tengan un valor únicamente si la columna originalmente nombrada como 'Current Status' es 'complete'

In [65]:
bad = data[(data['current_status_complete'] == 0) & ~(np.isnat(data['completed_date']))]
bad2 = data[(data['current_status_complete'] == 1) & (np.isnat(data['completed_date']))]
print(len(bad))
print(len(bad2))

114
0


Hay 114 filas que son inconsistentes. Nos quedamos únicamente con las consistentes:

In [66]:
data = data[(data['current_status_complete'] == 0) & (np.isnat(data['completed_date']))]

## Análisis y documentación de outliers

### tidf_compliance

La columna 'tidf_compliance' si bien luego de este pre-procesamiento de datos muestra 3 valores, es conveniente
dejar sentado que orignalmente esa columna posería una sola fila con un valor 'Y', otra con 'P' y el resto nulas.
En este laboratorio se asumió que los valores faltantes se corresponden con un valor negativo. En otras columnas donde ocurría algo similar, la cantidad de valores 'Y' y valores faltantes eran más balanceadas que este caso extremo.

## Comentarios finales

En el presente laboratorio se buscó realizar las limpiezas y conversiones de datos según el checklist sugerido en clase. Si bien el trabajo ha estado en nuestra consideración lejos de ser lo suficientemente exhaustivo como demandaría un caso real, lo que se ha buscado en el presente es mínimamente aplicar cada uno de los elementos del checklist en algunas columnas o valores.

En el caso de las conversiones de variables categóricas, la cantidad de columnas generadas consideramos que es demasiado grande y requeriría un análisis mas meticuloso para determinar si todos los valores convertidos en columnas son realmente necesarios. Hay casos en los que un puñado de valores posibles tienen una frecuencia alta, y luego hay cientos de valores con 1 sola ocurrencia. Estos ultimos podrian, por ejemplo, agruparse en una categoría "otros" para generar una sola columna.

No hemos podido almacenar en disco el dataframe resultante de todo este proceso de transformación, debido a su gran tamaño. Esto es por la conversión de las variables categóricas que se menciona en el párrafo anterior.

Por último, para que esta proceso sea completo, sería necesario analizar más casos de outliers.